Add opportunity to mixture word-vectors. Fix #1879

kakshay21 opened the pull request:
see https://github.com/RaRe-Technologies/gensim/issues/1879

- [x] Add initial benchmark

- [x] Remove `tf_idf` and use from gensim

- [x] Select one standard dataset and train the model

menshikh-iv commented:
@kakshay21 I don't see benchmark:

- you need to go through paper about models (like `Paragraph2vec`, `Sent2Vec`, etc) and look into `evaluation` section
- extract mentioned datasets and tasks
- compare document-embeddings methods with "mixture of word-embeddings" based on extracted datasets/tasks

kakshay21 commented:
Current Task Details:

1. Added sent2vec model.

2. Added doc2vec model. 

3. Added gensim API for the tfidf method.
 
TODO:

- [ ] Add benchmark.

- [ ] Implement the methods in gensim.

Evaluation Task (WIP):

- sent2vec evaluation task used cosine similarity with STS 2014 and SICK 2014 datasets.

- SIF evaluation task used similarity (SICK), entailment (SICK), sentiment (SST).

- doc2vec also seems to have similar task evaluation with sent2vec with additional tests with word2vec.

This is an initial mockup for API :
All mixture methods in `gensim.models.keyedvectors`
```

Smooth Inverse Frequency:

“””
Mixture of word-vectors to make sentence-vector.

Attains this by taking the average of word-vectors with the weight of words from the sentences. And then subtracting the projection of yielded vectors to their first principal component.

Parameters
------------
sentences: iterator of list of list of sentences
	Corpus of sentences.
W : np.array
	Vector embeddings of words from the pre-trained model.
a : float
	weighting parameter for best performance a lies between [10^-3,10^-4].

Returns
--------
embs : array of array of sentence embeddings
	 Sentence embeddings of each sentence in the same order as given.

“””


Simple Average:

“””
Mixture of word-vectors to make sentence-vector.

Attains this by simple averaging of all word-vectors.

Parameters
------------
sentences : iterator of list of list of sentences
	Corpus of sentences.
W : np.array
	Vector embeddings of words from the pre-trained model.

Returns
--------
embs : array of array of sentence embeddings
	 Sentence embeddings of each sentences in same order as given.

“””

TFIDF :

“””
Mixture of word-vectors to make sentence-vector.
 Attains this by averaging the word-vectors with multiplication of their TFIDF weights.

Parameters
------------
sentences : iterator of list of list of sentences
	Corpus of sentences.
W : np.array
	Vector embeddings of words from pre-trained model.

Returns
--------
embs : array of array of sentence embeddings
	 Sentence embeddings of each sentences in same order as given.

“””
```

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -55,8 +55,12 @@
 from __future__ import division  # py3 "true division"
+from gensim.models import TfidfModel
+from gensim.corpora import Dictionary
+from scipy.sparse.linalg import svds
```
We have our own decomposition, look at `gensim.models.lsimodel`

kakshay21 replied:
I just need [this](https://github.com/RaRe-Technologies/gensim/blob/49e6abdde6bca55b7a5b263c01ecdc94a75fb1ab/gensim/models/lsimodel.py#L293) step not sure if it is useful.

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+
+        Examples
+        --------
+        >>> sent = iter([['sample', 'sentence', 'cat', 'dog'], ['time', 'computers', 'expensive']])
```
`iter` isn't needed

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+        Attains this by simple averaging of all word-vectors.
+
+        Examples
+        --------
```
also, please write all imports directly in example (this should works if I copy-paste it to console)

menshikh-iv replied:
still not working (for all examples)

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+
+        Parameters
+        ------------
+        sent: iterator of list of tokens of sentences
```
incorrect type

menshikh-iv replied:
nitpick: missed space before `:` (here and everywhere)

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+
+        Returns
+        --------
+        sents_emd: class: `gensim.models.keyedvectors.Word2VecKeyedVectors`
```
this doesn't rendered correctly (please check rendered version always)

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+            Sentence embeddings of each sentence in the same order as given.
+
+        """
+        sents_emd = []
```
you should make embedding **for one** `sentence` (not for sequence of sentences), or support both variant

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+            new_sents_emd.append(s)
+        return new_sents_emd
+
+    def tf_idf_sent(self, sent):
```
IMO, better to use passed tfidf from user

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,134 @@ def get_keras_embedding(self, train_embeddings=False):
+            sents_emd.append(result)
+        return sents_emd
+
+    def smooth_inverse_frequency(self, sent, a=0.001):
```
maybe ask the user about dictionary too?

kakshay21 replied:
did you mean pre-trained model?

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,139 @@ def get_keras_embedding(self, train_embeddings=False):
+
+        Returns
+        -------
+        new_sents_emd : class : `gensim.models.keyedvectors.Word2VecKeyedVectors`
```
Incorrect type (here and everywhere for `Returns` section)

menshikh-iv commented on gensim/models/keyedvectors.py:
```diff
@@ -1063,6 +1067,139 @@ def get_keras_embedding(self, train_embeddings=False):
+
+        """
+        sents_emd = []
+        for s in sent:
```
if I pass one sentence - this didn't work

kakshay21 replied:
We talked about this since SIF and TFIDF need full corpus and to make it equivalent to all three mixture methods need full corpus, not just one sentence.

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "| S.No. | Model Name               | Pearson/Spearman/MSE on SICK | Mean SUBJ | Mean MR | TREC |\n",
```
Generate this table interactively (not manually)

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Download the pretrained model form [here](https://github.com/RaRe-Technologies/gensim-data/releases/tag/glove-wiki-gigaword-200)"
```
Use gensim-data api here (not manually download it)

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+    "from gensim.parsing.preprocessing import STOPWORDS\n",
+    "from scipy.sparse.linalg import svds\n",
+    "from scipy.spatial.distance import cosine\n",
+    "import numpy as np"
```
Missing description/motivation for notebook (what is it, why we do this)?

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
```
Cells was run chaotically (something not running at all, something running in strange order)

kakshay21 replied:
Less RAM, can't load all these model on a single run. I'll try this at the end.

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+    }
+   ],
+   "source": [
+    "for doc in lsi_corpus:\n",
```
???

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Implementation of mixture methods"
```
You already have this code in `keyedvectors.py`, please use it (instead of duplication)

menshikh-iv commented on docs/notebooks/word_to_sentence_embedding.ipynb:
```diff
@@ -0,0 +1,1264 @@
+    "import os\n",
+    "\n",
+    "\n",
+    "def load_data(loc='./data/'):\n",
```
and how I can retrieve this dataset? This will works only on your machine.
